pkg/doublewrite/doublewritebackend.go

```
@@ -0,0 +1,198 @@
+	crdBackend, _ := identitybackend.NewCRDBackend(c.CRDBackendConfiguration)
+	kvstoreBackend, _ := kvstoreallocator.NewKVStoreBackend(c.KVStoreBackendConfiguration)
+
+	log.WithFields(logrus.Fields{
+		"KVStoreBackendConfiguration.Suffix":   c.KVStoreBackendConfiguration.Suffix,
+		"KVStoreBackendConfiguration.Typ":      c.KVStoreBackendConfiguration.Typ.String(),
+		"KVStoreBackendConfiguration.BasePath": c.KVStoreBackendConfiguration.BasePath,
+		"readFromKVStore":                      c.ReadFromKVStore,
+	}).Debug("Creating the Double-Write backend")
+
+	return &doubleWriteBackend{crdBackend: crdBackend.(*identitybackend.CRDBackend), kvstoreBackend: kvstoreBackend, readFromKVStore: c.ReadFromKVStore}, nil
+}
+
+type DoubleWriteBackendConfiguration struct {
+	CRDBackendConfiguration     identitybackend.CRDBackendConfiguration
+	KVStoreBackendConfiguration kvstoreallocator.KVStoreBackendConfiguration
+	ReadFromKVStore             bool
+}
+
+type doubleWriteBackend struct {
+	crdBackend      *identitybackend.CRDBackend
+	kvstoreBackend  *kvstoreallocator.KVStoreBackend
+	readFromKVStore bool
+}
+
+func (d *doubleWriteBackend) DeleteAllKeys(ctx context.Context) {
+	d.crdBackend.DeleteAllKeys(ctx)
+	d.kvstoreBackend.DeleteAllKeys(ctx)
+}
+
+func (d *doubleWriteBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) (allocator.AllocatorKey, error) {
+	crdKey, crdErr := d.crdBackend.AllocateID(ctx, id, key)
+	if crdErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr).Error("CRD backend failed to allocate identity")
+	}
+	kvStoreKey, kvStoreErr := d.kvstoreBackend.AllocateID(ctx, id, key)
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity")
+	}
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
```

Why is the error returned for a single backend only? I'd personally assume that we want to return it from both for all "write" operations, to ensure that we subsequently perform a retry even if only one failed for consistency. Am I missing anything? 

Another aspect to consider is how to ensure that these operations are atomic. Let's say that for instance the allocation succeeded in CRD mode, but then failed in kvstore mode. My understanding is that we should rollback the CRD allocation to remove the spurious CiliumIdentity object.

---

So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit. 

You would need a system to rollback to a previous state if one of the writes failed: for object creations that would probably be possible with careful deletes but for object updates you would need to store the previous state in memory and also account for race conditions due to concurrent updates... IMO that's too complex, especially for a backend which is supposed to be used only for a short period of time.

Moreover, the current code already handles failed writes to the "primary" backend (i.e. the one used for reads) and there is alerting which helps with detecting discrepancies (in case a write to the "primary" backend succeeded but failed for the "secondary" backend).

---

> So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit.

Yeah, fair enough. I agree that it makes sense to skip that considering that it may not be possible to actually revert the previous changes, or lead to additional inconsistencies.

I still wonder whether it would be better to return an error if any of the two backend operations failed though, as that would allow the caller to detect that something went wrong, and perform a retry. My assumption is that it should help to automatically resolve temporary errors (e.g., kvstore blips, or creation conflicts), which may otherwise be retried only during a subsequent agent restart. WDYT?

---

I looked into it a bit more and I could technically refactor the code to return errors with something like this:
```diff
@@ -122,10 +126,12 @@ func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key al
        crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
        if crdErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
+               return crdErr
        }
        kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
        if kvStoreErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
+               return kvStoreErr
        }
```
However, this adds a small complication: now if a CRD update fails, the kvstore update will never go through. My initial idea was that in `doublewrite-readkvstore` mode, the kvstore writes should be the "primary" ones and CRD writes are made on a best-effort basis, trying to not interfere with kvstore operations (and vice-versa in `doublewrite-readcrd` mode). This way the `doublewrite-readkvstore` functions in the same way as the regular `kvstore` mode but with best-effort CRD writes in parallel.

I could refactor the code even more to take this into account:
```go
func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
	if d.readFromKVStore {
		// If we are reading from the kvstore, we perform the kvstore update first
		kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
		if kvStoreErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
			return kvStoreErr
		}
		crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
		if crdErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
			return crdErr
		}
	}
	// If we are reading from CRDs, we perform the CRD update first
	crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
	if crdErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
		return crdErr
	}
	kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
	if kvStoreErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
		return kvStoreErr
	}
	return nil
}
```
But this makes the code more convoluted (I'll need to write the same logic for all write methods basically). What's your take on this?

---

> I looked into it a bit more and I could technically refactor the code to return errors with something like this

Yeah, that is what I was thinking about when I wrote the above comment, so that the operation would be retried regardless of which backend failed. My main concern with only returning errors for the primary backend is that missing identities on the secondary one would cause divergences, that would be possibly fixed on agent restart only.

> However, this adds a small complication: now if a CRD update fails, the kvstore update will never go through.

Yes, but the operation would be retried, and eventually both backends should succeed. I agree that a persistent failure of one would also break the other, but that seems a pretty catastrophic situation to me.

> I could refactor the code even more to take this into account:

I agree that this looks pretty convoluted, and I don't think it would help much either, as, at least for a few methods, the caller would just retry with a different ID. Hence, there's not much point in ensuring that the primary update goes through first if the identity is discarded anyways in case of errors. 

In any case, I'd like to also get a second opinion of someone with more context on the internals of the allocator logic (@squeed maybe?).

---

Because migration is expected to be reversible (as outlined in the original proposal), we must always write to both datastores, as @giorio94 observes.

This is especially critical given that different agents may have a different understanding of "primary". As configuration changes are rolled out, *if* we were to ignore secondary-side failures, we could have pretty serious divergence.

Given that most migrations are probably to the CRD store, and if the k8s apiserver goes down, nothing is happening anyways, I believe the right pattern is to write to CRD first, then kvstore. But, if kvstore fails, we must revert our CRD change and return failure.

There are still some nasty corner cases around interleaving failures between doublewrite and non-doublewrite agents, but I think those are very unlikely.

---

Makes sense, I wrote some code to handle this for the allocation functions: https://github.com/cilium/cilium/pull/31920/files#diff-ce8e3e1201058922bca6b6b8fd0ffc3a9ac595a34621246e05daf000f343bcb5R80-R100 
LMK what you think.

And also added a test: https://github.com/cilium/cilium/pull/31920/files#diff-08f67e47fb93b88a978cbf9b31e32f4af4b01ac36b1a78b9890201cf654b8d6aR89-R109

I didn't add the same logic for the `UpdateKey` / `AcquireReference` functions because rolling back the CRD update is not that important there: these functions basically update the heartbeat timestamp on the CRD object, so in the worst case, the GC of the identity will be delayed.